src/test/java/rx/observables/AsyncOnSubscribeTest.java

```diff
@@ -0,0 +1,408 @@
+import rx.functions.Action1;
+import rx.functions.Action2;
+import rx.functions.Func0;
+import rx.functions.Func3;
+import rx.internal.util.RxRingBuffer;
+import rx.observers.TestSubscriber;
+import rx.schedulers.TestScheduler;
+
+@RunWith(MockitoJUnitRunner.class)
+public class AsyncOnSubscribeTest {
+
+    @Mock
+    public Observer<Integer> o;
+
+    TestSubscriber<Integer> subscriber;
+    
+    @Before
+    public void setup() {
+        subscriber = new TestSubscriber<Integer>(o);
+    }
+    
+    @Test
+    public void testSerializesConcurrentObservables() throws InterruptedException {
+        final TestScheduler scheduler = new TestScheduler();
+        OnSubscribe<Integer> os = AsyncOnSubscribe.createStateful(new Func0<Integer>(){
+            @Override
+            public Integer call() {
+                return 1;
+            }}, 
+        new Func3<Integer, Long, Observer<Observable<? extends Integer>>, Integer>(){
+            @Override
+            public Integer call(Integer state, Long requested, Observer<Observable<? extends Integer>> observer) {
+                if (state == 1) {
+                    Observable<Integer> o1 = Observable.just(1, 2, 3, 4)
+                            .delay(100, TimeUnit.MILLISECONDS, scheduler);
+                    observer.onNext(o1);
+                }
+                else if (state == 2) {
+                    Observable<Integer> o = Observable.just(5, 6, 7, 8);
+                    observer.onNext(o);
+                }
+                else 
+                    observer.onCompleted();
+                return state + 1;
+            }});
+        // initial request emits [[1, 2, 3, 4]] on delay
+        Observable.create(os).subscribe(subscriber);
+        // next request emits [[5, 6, 7, 8]] firing immediately
+        subscriber.requestMore(2); 
+        // triggers delayed observable
+        scheduler.advanceTimeBy(100, TimeUnit.MILLISECONDS);
+        // final request completes
+        subscriber.requestMore(3); 
+        subscriber.awaitTerminalEventAndUnsubscribeOnTimeout(100, TimeUnit.MILLISECONDS);
+        subscriber.assertNoErrors();
+        subscriber.assertReceivedOnNext(Arrays.asList(new Integer[] {1, 2, 3, 4, 5, 6, 7, 8}));
+        subscriber.assertCompleted();
+    }
+    
+    @Test
+    public void testSubscribedByBufferingOperator() {
+        final TestScheduler scheduler = new TestScheduler();
+        OnSubscribe<Integer> os = AsyncOnSubscribe.createStateless(
+        new Action2<Long, Observer<Observable<? extends Integer>>>(){
+            @Override
+            public void call(Long requested, Observer<Observable<? extends Integer>> observer) {
+                observer.onNext(Observable.range(1, requested.intValue()));
+            }});
+        Observable.create(os).observeOn(scheduler).subscribe(subscriber);
+        scheduler.advanceTimeBy(10, TimeUnit.DAYS);
+        subscriber.assertNoErrors();
+        subscriber.assertValueCount(RxRingBuffer.SIZE);
```

Comment:
I don't understand how `subscriber` receives only `RxRingBuffer.SIZE` values and `observeOn` doesn't request more if `subscriber` appears to be requesting Long.MAX_VALUE?